### PR TITLE
WIP: SavingCallback

### PR DESCRIPTION
**Pull request by ranocha**
This came up in https://github.com/JuliaDiffEq/DifferentialEquations.jl/issues/202.

It is a basic attempt of an implementation. It allows to save arbitrary values `save_func(t, u, integrator)` and the corresponding time `t` in a struct `SavedValues`.

However, I'm not fully satisfied with the approach, since the struct `SavedValues` has to be created at first, knowing the return type of `save_func(t, u, integrator)`. If the callback could be created after the initialisation of the integrator, this return type could be set automatically. 

Does anyone have an idea to improve this approach?

### CODE REVIEW

**Comment by ranocha** [comment]
Travis fails in 
```
Domain tests: Error During Test
  Got an exception of type LoadError outside of a @test
  LoadError: MethodError: no method matching jacobian!(::DiffBase.MutableDiffResult{1,Array{Float64,1},Tuple{Array{Float64,2}}}, ::DiffEqCallbacks.##6#10{#g,Array{Float64,1}}, ::Array{Float64,1}, ::Array{Float64,1}, ::ForwardDiff.JacobianConfig{ForwardDiff.Tag{Void,0xb046287d533b082d},Float64,1,Tuple{Array{ForwardDiff.Dual{ForwardDiff.Tag{Void,0xb046287d533b082d},Float64,1},1},Array{ForwardDiff.Dual{ForwardDiff.Tag{Void,0xb046287d533b082d},Float64,1},1}}})
```
There are no errors on my machine. Does anybody know what's happening?

**Comment by ranocha** [comment]
This is an issue with the new version `v0.6.0` of ForwardDiff.jl, see https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/200.

**Comment by ranocha** [comment]
The coverage decreased slightly due to docstrings in the new code.

File: src/DiffEqCallbacks.jl [hunk 1]
```diff
@@ -2,14 +2,17 @@ __precompile__()
 
 module DiffEqCallbacks
 
-  using DiffEqBase, NLsolve, ForwardDiff
-  import DiffBase
+  using DiffEqBase, NLsolve, ForwardDiff, RecursiveArrayTools, DataStructures
```
Comment (ChrisRackauckas) [inline src/DiffEqCallbacks.jl]: Need to add these to REQUIRE

File: src/saving.jl [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+"""
+    SavedValues(tType::DataType, savevalType::DataType)
+
+Return `SavedValues{tType, savevalType}` with empty storage vectors.
+"""
+function SavedValues(tType::DataType, savevalType::DataType)
+    SavedValues{tType, savevalType}(Vector{tType}(), Vector{savevalType}())
+end
+
+function show(io::IO, saved_values::SavedValues)
```
Comment (ChrisRackauckas) [inline src/saving.jl]: Base.show

File: src/DiffEqCallbacks.jl [hunk 3]
```diff
@@ -2,14 +2,17 @@ __precompile__()
...
+  import DiffResults
 
-  import OrdinaryDiffEq: fix_dt_at_bounds!, modify_dt_for_tstops!
+  import OrdinaryDiffEq: fix_dt_at_bounds!, modify_dt_for_tstops!, ode_addsteps!, ode_interpolant
+
+  import Base: show
```
Comment (ChrisRackauckas) [inline src/DiffEqCallbacks.jl]: Could you instead explicitly use `Base.show`? I find that less error prone.

**Comment by ChrisRackauckas** [comment]
Coverage decrease is fine.

Could you also add a quick blurb to the README?

I am also not sure how the types could be automatically deduced.

>If the callback could be created after the initialisation of the integrator, this return type could be set automatically.

That sounds like it could be possible in the future via the `initialize!` part of the callback, but would require that we don't use the explicit function typing in the callback, which to be type-stable we would then need to change using some kind of `FunctionWrapper` and then no longer use the compile-time recursion tricks. Let's not handle that for now.

**Comment by ChrisRackauckas** [comment]
Will this still work with the new version bounds?

**Comment by ranocha** [comment]
Since I've updated the manifold projection callback, it requires (at least) ForwardDiff.jl `v0.6.0`, so it won't work with OrdinaryDiffEq.jl if there's an upper version bound `v0.5.0` on ForwardDiff.jl.

**Comment by ranocha** [comment]
Travis and AppVeyor use version `v2.4.0` of OrdinaryDiffEq.jl, so they fail.

**Comment by ranocha** [comment]
Now, this will work with ForwardDiff `v0.5.0` and excludes `v0.6.0`. I'll open an issue for the update to `v0.6.0`.

**Pull request merged**